### PR TITLE
[UR][L0] Add several fixes to L0 adapter for test-usm

### PR DESCRIPTION
**Pull request by jandres742**
intel/llvm testing: https://github.com/intel/llvm/pull/11980

### CODE REVIEW

File: source/adapters/level_zero/usm.cpp [hunk 1]
```diff
@@ -597,6 +617,40 @@ UR_APIEXPORT ur_result_t UR_APICALL urUSMGetMemAllocInfo(
...
+
+    for (auto &Pool : Context->UsmPoolHandles) {
+      if (SearchMatchingPool(Pool->DeviceMemPools, UMFPool)) {
+        return ReturnValue(Pool);
+      }
+      if (SearchMatchingPool(Pool->SharedMemPools, UMFPool)) {
+        return ReturnValue(Pool);
+      }
+      if (Pool->HostMemPool.get() == UMFPool) {
+        return ReturnValue(Pool->HostMemPool.get());
```
Comment (kswiecicki) [inline source/adapters/level_zero/usm.cpp:648]: This should also return `ReturnValue(Pool);`

File: source/adapters/level_zero/usm.cpp [hunk 2]
```diff
@@ -829,6 +884,10 @@ UR_APIEXPORT ur_result_t UR_APICALL urUSMPoolCreate(
   try {
     *Pool = reinterpret_cast<ur_usm_pool_handle_t>(
         new ur_usm_pool_handle_t_(Context, PoolDesc));
+
+    std::shared_lock<ur_shared_mutex> ContextLock(Context->Mutex);
+    Context->UsmPoolHandles.push_back(*Pool);
```
Comment (kswiecicki) [inline source/adapters/level_zero/usm.cpp:889]: `Context->UsmPoolHandles` should also be updated in `urUSMPoolRelease`.

File: source/adapters/level_zero/usm.cpp (hunk 1 above)
Comment (jandres742) [reply source/adapters/level_zero/usm.cpp:648]: thanks @kswiecicki! fixed.

File: source/adapters/level_zero/usm.cpp (hunk 2 above)
Comment (jandres742) [reply source/adapters/level_zero/usm.cpp:889]: thanks @kswiecicki. missed it. fixed.

**Comment by jandres742** [comment]
There are some tests failing because we need to have UMF tracking. So we need to have this one merged https://github.com/oneapi-src/unified-runtime/pull/1071

**Comment by fabiomestre** [comment]
I have updated the target branch of this PR from the `adapters` branch to the `main` branch.
Development in UR is moving back to `main`. The `adapters` branch will soon be deleted.

**Comment by kbenzie** [comment]
> There are some tests failing because we need to have UMF tracking. So we need to have this one merged #1071

I'm merged #1071 now so will rebase this on top of the main branch.

**Comment by kbenzie** [comment]
I updated https://github.com/intel/llvm/pull/11980 to point at the latest commit on this branch and its caused [121 failures](https://github.com/intel/llvm/actions/runs/7183686782/job/19563679849?pr=11980). @nrspruit I think I've taken this one as far as I can. I'll need to hand that over to you.

**Comment by nrspruit** [comment]
> I updated [intel/llvm#11980](https://github.com/intel/llvm/pull/11980) to point at the latest commit on this branch and its caused [121 failures](https://github.com/intel/llvm/actions/runs/7183686782/job/19563679849?pr=11980). @nrspruit I think I've taken this one as far as I can. I'll need to hand that over to you.

Thanks @kbenzie , I will be taking a look tomorrow to determine what is causing the failures.

File: source/adapters/level_zero/usm.cpp [hunk 3]
```diff
@@ -187,8 +187,15 @@ static ur_result_t USMDeviceAllocImpl(void **ResultPtr,
     ZeDesc.pNext = &RelaxedDesc;
   }
 
-  ZE2UR_CALL(zeMemAllocDevice, (Context->ZeContext, &ZeDesc, Size, Alignment,
```
Comment (aarongreig) [inline source/adapters/level_zero/usm.cpp:190]: Looks like this macro was needed for the l0 api tracing enabled with `UR_L0_DEBUG`, that tracing it necessary for a lot of the sycl-e2e tests (they check for certain l0 api calls in the trace output). We should make these kinds of changes like this
```
ze_result_t ZeResult = ZE_CALL_NOCHECK(zeMemAllocDevice, (Context->ZeContext,
              &ZeDesc, Size, Alignment, Device->ZeDevice, ResultPtr));
```
instead

**Comment by kbenzie** [comment]
Looks like the SYCL E2E failures have been resolved so I'll merge this.

**Pull request merged**